crypto/rsa/rsa_gen.c

```diff
@@ -234,15 +233,11 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, int primes, BIGNUM *e_value,
          * distinguish a multi-prime private key by using the modulus
          * in a certificate.
          */
-        nlb = (nl + 7) / 8;
-        bin = OPENSSL_malloc(nlb);
-        if (bin == NULL)
+        if (!BN_rshift(r2, r1, nl - 4))
             goto err;
-        (void)BN_bn2bin(r1, bin);
-        bint = bin[0] >> 4;
-        OPENSSL_free(bin);
+        bitst = BN_get_word(r2);
 
-        if (nl != bitse || (bint < 0x9 && i == primes - 1)) {
+        if (nl != bitse || bitst < 0x9) {
```

Comment:
... and `if (bitst < 0x9 || bitst > 0xF)` here, then you will probably not need `nl` any more.

Reply:
Seems bitst could never be bigger than 0xF

Reply:
in the first iteration, yes,
but later that depends in you adjust value.

Reply:
BTW: I do not understand why you need `bitse` at all, that value should
match `bits` here?

Reply:
but bitst is limited to 4 bits only, how could it exceed 1111 ?

Reply:
> I do not understand why you need bitse at all, that value should
match bits here?

The idea is: bitse is used to present the expected length of the product after each prime is generated and is multiplied to `rsa->n`. And if the actual length is not the same as what is expected, re-generate the last prime.

Reply:
Hmm, yes, I see.
So the bitst < 0x9 is definitely needed in the last iteration.
But it may be possible that it works more smoothly when
all intermediate primes use the same check.